d2l-visible-on-ancestor-behavior.html

```diff
@@ -0,0 +1,157 @@
+
+	/** @polymerBehavior */
+	D2L.PolymerBehaviors.VisibleOnAncestorBehavior = {
+
+		properties: {
+
+			/**
+			 * Whether the element is only visible when hovered or focus is within ancestor with 'd2l-visible-on-ancestor-target'.
+			 */
+			visibleOnAncestor: {
+				type: Boolean,
+				observer: '__voaHandleVisibleOnAncestor',
+				reflectToAttribute: true
+			}
+
+		},
+
+		detached: function() {
+			this.__voaUninit();
+		},
+
+		__voaHandleHideEnd: function(e) {
+			if (e.propertyName !== 'transform') {
+				return;
+			}
+			this.removeEventListener('transitionend', this.__voaHandleHideEnd);
+			fastdom.mutate(function() {
+				this.setAttribute('d2l-visible-on-ancestor-hide', 'd2l-visible-on-ancestor-hide');
+			}.bind(this));
+		},
+
+		__voaHandleFocus: function() {
+			this.__voaFocusIn = true;
+			this.__voaShow();
+		},
+
+		__voaHandleBlur: function(e) {
+			if (D2L.Dom.isComposedAncestor(this.__voaTarget, e.relatedTarget)) {
+				return;
+			}
+			this.__voaFocusIn = false;
+			this.__voaHide();
+		},
+
+		__voaHandleMouseEnter: function() {
+			this.__voaMouseOver = true;
+			this.__voaShow();
+		},
+
+		__voaHandleMouseLeave: function() {
+			this.__voaMouseOver = false;
+			this.__voaHide();
+		},
+
+		__voaHandleVisibleOnAncestor: function(newValue, oldValue) {
+			if (newValue) {
+				this.__voaInit();
+			} else if (oldValue) {
+				this.__voaUninit();
+			}
+		},
+
+		__voaHide: function() {
+			if (this.__voaFocusIn || this.__voaMouseOver) return;
+			fastdom.mutate(function() {
+				this.addEventListener('transitionend', this.__voaHandleHideEnd);
+				this.setAttribute('d2l-visible-on-ancestor-transition', 'd2l-visible-on-ancestor-transition');
+				this.setAttribute('d2l-visible-on-ancestor-hide', 'd2l-visible-on-ancestor-hide');
+				this.removeAttribute('d2l-visible-on-ancestor-show');
+			}.bind(this));
+		},
+
+		__voaInit: function() {
+
+			this.__voaTarget = D2L.Dom.findComposedAncestor(this, function(node) {
+				if (!node || node.nodeType !== 1) return false;
+				return (node.classList.contains('d2l-visible-on-ancestor-target'));
+			});
+			if (!this.__voaTarget) return;
+
+			this.__voaHandleBlur = this.__voaHandleBlur.bind(this);
+			this.__voaHandleFocus = this.__voaHandleFocus.bind(this);
+			this.__voaHandleMouseEnter = this.__voaHandleMouseEnter.bind(this);
+			this.__voaHandleMouseLeave = this.__voaHandleMouseLeave.bind(this);
+			this.__voaHandleHideEnd = this.__voaHandleHideEnd.bind(this);
+
+			this.__voaTarget.addEventListener('focus', this.__voaHandleFocus, true);
+			this.__voaTarget.addEventListener('blur', this.__voaHandleBlur, true);
+			this.__voaTarget.addEventListener('mouseenter', this.__voaHandleMouseEnter);
+			this.__voaTarget.addEventListener('mouseleave', this.__voaHandleMouseLeave);
+
+			fastdom.mutate(function() {
+				this.setAttribute('d2l-visible-on-ancestor-hide', 'd2l-visible-on-ancestor-hide');
+			}.bind(this));
+
+			fastdom.measure(function() {
```

Comment:
The expectation is that we'll have translucent buttons and dropdown openers, so the end-state of opacity needs to be whatever value it's using.